src/lib/tls/msg_client_hello.cpp

```diff
@@ -478,7 +478,19 @@ Client_Hello_12::Client_Hello_12(Handshake_IO& io,
       m_data->extensions().add(new Certificate_Status_Request({}, {}));
    }
 
-   auto supported_groups = std::make_unique<Supported_Groups>(policy.key_exchange_groups());
+   // RFC 7919 3.
+   //    A client that offers a group MUST be able and willing to perform a DH
+   //    key exchange using that group.
+   //
+   // We don't support hybrid key exchange in TLS 1.2
+   const std::vector<Group_Params> kex_groups = policy.key_exchange_groups();
+   std::vector<Group_Params> compatible_kex_groups;
+   std::copy_if(kex_groups.begin(), kex_groups.end(), std::back_inserter(compatible_kex_groups), [](const auto group) {
+      return is_ecdh(group) || is_dh(group) || is_x25519(group);
+   });
```

Comment:
I would suggest pulling these two common blocks into a utility function in this file, something like

`std::unique_ptr<Supported_Groups> make_tls12_supported_groups(const Policy& p)`